tensorflow/compiler/xla/service/slice_delaying.cc

```diff
@@ -0,0 +1,306 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/slice_delaying.h"
+#include <algorithm>
+#include <utility>
+
+namespace xla {
+
+bool SliceLess(const HloInstruction* lhs, const HloInstruction* rhs) {
+  int64 l = 0;
+  int64 r = 0;
+  for (int i = 0; i < lhs->slice_starts().size(); ++i) {
+    l = lhs->slice_starts(i);
+    r = rhs->slice_starts(i);
+    if (l < r)
+      return true;
+    if (l > r)
+      return false;
+  }
+  return false;
+}
+
+bool SliceDelaying::CheckSlice(const HloInstruction* inst) {
+  if (inst->opcode() != HloOpcode::kSlice)
+    return false;
+  for (int64 i : inst->slice_strides())
+    if (i != 1)
+      return false;
+  return true;
+}
+
+// Only support split one dimension
+int SliceDelaying::SplitDim(const HloInstruction* inst,
+                            const HloInstruction* slice) {
+  const Shape shape = inst->shape();
+  const Shape slice_shape = slice->shape();
+  int slice_dimension = -1;
+  if (ShapeUtil::TrueRank(shape) != ShapeUtil::TrueRank(slice_shape))
+    return -1;
+  for (int i = 0; i < ShapeUtil::TrueRank(shape); ++i) {
+    if (shape.dimensions(i) == slice_shape.dimensions(i))
+      continue;
+    if (slice_dimension != -1)
+      return -1;
+    slice_dimension = i;
+  }
+  return slice_dimension;
+}
+
+bool SliceDelaying::CheckContinuous(
```

Comment:
Again needs a comment and probably a better function name.  Or we can just inline it into the one caller as

```
  absl::c_all_of(slices, [&](const HloInstruction* instr) { instr->slice_limits(dim) == slices[0]->slice_limits(dim); });
```